vms/platformvm/block/builder/builder.go

```diff
@@ -86,81 +72,55 @@ type builder struct {
 
 	txExecutorBackend *txexecutor.Backend
 	blkManager        blockexecutor.Manager
-
-	// resetTimer is used to signal that the block builder timer should update
-	// when it will trigger building of a block.
-	resetTimer chan struct{}
-	closed     chan struct{}
-	closeOnce  sync.Once
-	notify     func() // Notifies when a block is built
 }
 
 func New(
 	mempool mempool.Mempool[*txs.Tx],
 	txExecutorBackend *txexecutor.Backend,
 	blkManager blockexecutor.Manager,
-	notify func(),
 ) Builder {
 	return &builder{
-		notify:            notify,
 		Mempool:           mempool,
 		txExecutorBackend: txExecutorBackend,
 		blkManager:        blkManager,
-		resetTimer:        make(chan struct{}, 1),
-		closed:            make(chan struct{}),
 	}
 }
 
-func (b *builder) StartBlockTimer() {
-	go func() {
-		timer := time.NewTimer(0)
-		defer timer.Stop()
-
-		for {
-			// Invariant: The [timer] is not stopped.
-			select {
-			case <-timer.C:
-			case <-b.resetTimer:
-				if !timer.Stop() {
-					<-timer.C
-				}
-			case <-b.closed:
-				return
-			}
-
-			// Note: Because the context lock is not held here, it is possible
-			// that [ShutdownBlockTimer] is called concurrently with this
-			// execution.
-			for {
-				duration, err := b.durationToSleep()
-				if err != nil {
-					b.txExecutorBackend.Ctx.Log.Error("block builder encountered a fatal error",
-						zap.Error(err),
-					)
-					return
-				}
-
-				if duration > 0 {
-					timer.Reset(duration)
-					break
-				}
-
-				// Block needs to be issued to advance time.
-				b.notify()
-
-				// Invariant: ResetBlockTimer is guaranteed to be called after
-				// [durationToSleep] returns a value <= 0. This is because we
-				// are guaranteed to attempt to build block. After building a
-				// valid block, the chain will have its preference updated which
-				// may change the duration to sleep and trigger a timer reset.
-				select {
-				case <-b.resetTimer:
-				case <-b.closed:
-					return
-				}
-			}
+func (b *builder) WaitForEvent(ctx context.Context) (common.Message, error) {
+	for {
+		if err := ctx.Err(); err != nil {
+			return 0, err
+		}
+
+		duration, err := b.durationToSleep()
+		if err != nil {
+			b.txExecutorBackend.Ctx.Log.Error("block builder failed to calculate next staker change time",
+				zap.Error(err),
+			)
+			return 0, err
+		}
+		if duration <= 0 {
+			// The next staker change is ready to be performed.
+			return common.PendingTxs, nil
+		}
+
+		// Wait for a transaction in the mempool until there is a next staker
+		// change ready to be performed.
+		newCtx, cancel := context.WithTimeout(ctx, duration)
+		msg, err := b.Mempool.WaitForEvent(newCtx)
+		cancel()
+
+		switch {
+		case err == nil:
+			return msg, nil
+		case errors.Is(err, context.DeadlineExceeded):
+			continue // Recheck the staker change time before returning
```

Comment:
why do we re-check instead of returning pending?

Reply:
I did this to differentiate between the parent context and the child context reporting `DeadlineExceeded`.

Reply:
I suppose it may be better to just check `ctx.Err() != nil`?